samples: bluetooth: ipsp: support ipsp with disco_l475_iot1

cpriouzeau opened the pull request:
To support the bluetooth ipsp sample on STM disco_l475_iot1 board:
- Modify disco_l475_iot1 configuration for bluetooth support.
- Force usage of MSI clock as SYSCLK_SRC on bluetooth usage.
- Adapt hci driver over spi for BlueNRG-MS: need to read until IRQ pin goes low
- Adapt hci driver over spi: put BlueNRG-MS in controller mode
Rebased on master branch + depends on patch from #8239 (bluetooth quirks)

erwango commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -383,7 +424,7 @@ static int bt_spi_open(void)
 
 	/* Configure IRQ pin and the IRQ call-back/handler */
-	gpio_pin_configure(irq_dev, GPIO_IRQ_PIN,
+	gpio_pin_configure(irq_dev, GPIO_IRQ_PIN, GPIO_PUD_PULL_DOWN |
```
I think we should put this under CONFIG_BT_SPI_BLUENRG flag.

cpriouzeau replied:
DONE, available with new update

erwango commented on boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig:
```diff
@@ -22,6 +22,7 @@ CONFIG_CLOCK_CONTROL=y
 # Default: PLL driven by HSI
 CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
 CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
+#CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI=y
```
Remove this line (same as the one below)

cpriouzeau replied:
DONE, available with new update

erwango commented on boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig:
```diff
@@ -54,8 +55,15 @@ CONFIG_BT_SPI_CHIP_SELECT_DEV_NAME="GPIOD"
 CONFIG_HTS221_TRIGGER_NONE=y
+
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_BT_SPI_MAX_CLK_FREQ=2000000
```
Move this one up with CONFIG_BT_SPI_xxx friends

cpriouzeau replied:
DONE, available with new update

erwango commented on boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig:
```diff
@@ -54,8 +55,15 @@ CONFIG_BT_SPI_CHIP_SELECT_DEV_NAME="GPIOD"
 CONFIG_LIS3MDL_TRIGGER_NONE=y
 CONFIG_HTS221_TRIGGER_NONE=y
+
+CONFIG_UART_INTERRUPT_DRIVEN=y
```
group this one with CONFIG_SERIAL

cpriouzeau replied:
DONE, available with new update

erwango reviewed (changes requested):
Some comments otherwise LGTM

mike-scott commented:
Also the default settings for some of the buffers / HCI commands were not high enough to run the client.  I ended up throwing these settings at it to get it working w/o complaints:
CONFIG_BT_HCI_CMD_COUNT=10
CONFIG_BT_RX_BUF_COUNT=20
CONFIG_BT_RX_BUF_LEN=128

erwango commented:
@mike-scott , thanks for having had a try !

> This is probably a result of CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI=y .. is this expected?

To explain this change: we've been forced to use MSI because it appeared using HSI driving PLL was not accurate enought, and this setting was known working.
This being said, we didn't expect any side effect on time counting, I'll have a look.

Vudentz commented on samples/bluetooth/ipsp/prj_zep1656_disco_l475_iot1.conf:
```diff
@@ -0,0 +1,32 @@
+# CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL is not set
+CONFIG_CLOCK_STM32_SYSCLK_SRC_MSI=y
```
I wonder why this is not in the board configuration? Everything else seems to be the same as in prj.conf so if the board sets it we wouldn't need a specific prj.conf.

cpriouzeau replied:
In general, we don't need to use this clock, for example with beacon sample it's not mandatory to change the clock but with ipsp sample we have seen some data lost during bluetooth communication and is why we have added this configuration specifically for this sample.

mike-scott commented:
@cpriouzeau I tried setting up the defaults for:
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
CONFIG_CLOCK_STM32_PLL_SRC_HSI
in a slightly different way so that the prj.conf files wouldn't need to be altered.  I used the logic that we need to select CLOCK_STM32_SYSCLK_SRC_MSI if BT is enabled.

Here's the diff from your 4th patch (I removed the 5th patch so I could test).
https://hastebin.com/maweridifi.diff

This let me build samples/bluetooth/ipsp using the normal prj.conf as well as lwm2m-client with only the following minor additions to overlay-bt.conf:
```
CONFIG_BT_HCI_CMD_COUNT=10
CONFIG_BT_RX_BUF_COUNT=20
CONFIG_BT_RX_BUF_LEN=128
```

Also you may note that I moved CONFIG_SPI along with all of the CONFIG_BT_* settings into the Kconfig.defconfig under a check to see if CONFIG_BT was enabled.  This reduced the Kconfig warning spam by a great deal when building samples that don't need bluetooth.

I did note that when building lwm2m_client, I saw the following warning over and over:
```
In file included from zephyr/include/arch/arm/cortex_m/scripts/linker.ld:19:0,
                 from zephyr/arch/arm/soc/st_stm32/stm32l4/linker.ld:9:
zephyr/samples/net/lwm2m_client/build/zephyr/include/generated/generated_dts_board.h:438:0: warning: "CONFIG_BT_SPI_MAX_CLK_FREQ" redefined
 #define CONFIG_BT_SPI_MAX_CLK_FREQ  ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_SPI_MAX_FREQUENCY
 
In file included from zephyr/include/arch/arm/cortex_m/scripts/linker.ld:17:0,
                 from zephyr/arch/arm/soc/st_stm32/stm32l4/linker.ld:9:
zephyr/samples/net/lwm2m_client/build/zephyr/include/generated/autoconf.h:94:0: note: this is the location of the previous definition
 #define CONFIG_BT_SPI_MAX_CLK_FREQ 2000000
```

Question:
Is CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL the default SYSCLK on all STM32 hardware in Zephyr?  I did a quick search in boards/ and found that it was selected by MANY of the STM32 boards (if not all).  What are the exception cases?

cpriouzeau commented:
@mike-scott , For CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL clock, is the clock use by default on the majority of STM32 hardware but following use-cases you can use the other clock available like HSE, MSI. 

For he warning and configuration, on the next patchset (for this pull request), I will provide a rework of this part.

erwango commented:
> @mike-scott , For CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL clock, is the clock use by default on the majority of STM32 hardware but following use-cases you can use the other clock available like HSE, MSI.

For this specific BLE case, it appeared that using PLL from HSI was not accurate enough. Then PLL from HSE could not be used as it requires HSE which is not present on board. Last, PLL from MSI could not be used either since in that case MSI should use a LSE for auto calibration, which is not there neither.

erwango commented:
@cpriouzeau can you rebase and use #7350, then I think we'll be good.
Thanks

erwango commented:
@galak, can you have a look?

mike-scott commented on boards/arm/disco_l475_iot1/disco_l475_iot1_defconfig:
```diff
@@ -46,9 +39,12 @@ CONFIG_I2C=y
 # sensors configuration
 CONFIG_LIS3MDL_TRIGGER_NONE=y
 CONFIG_HTS221_TRIGGER_NONE=y
+
```
Are these extra lines needed?

cpriouzeau replied:
No , I will remove it.

erwango commented:
@cpriouzeau, if you rebase on #7538, we can get rid of CONFIG_BT_SPI=y

mike-scott reviewed (approved):
LGTM

cpriouzeau committed:
boards: disco_l475_iot1: update BT configuration

- update BT configuration
- adapt configuration of SPI3 for BT usage

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>

cpriouzeau commented:
@erwango rebased and CONFIG_BT_SPI put on kconfig.defconfig

galak commented:
Looks good to me, but should get @jhedberg approval on the bluetooth driver changes.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -670,6 +674,30 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(0xFC0C, sizeof(uint8_t)*(3));
```
Please create a #define for the 0xfc0c, constructed with BT_OP(BT_OGF_VS, 0x0003). Also, `sizeof(uint8_t)*(3)` seems like a really complicated way of saying just `3` :)  (not to mention that we use the Zephyr int types instead of stdint)

jhedberg replied:
An even better option than just '3' would be if you create a parameter struct consisting of three descriptively named u8_t member variables. Then you can do something like:

```
struct vendor_param {
        u8_t foo;
        u8_t bar;
        u8_t baz;
};
...
static int hci_send_aci_config_data_controller_mode(void)
{
        struct vendor_param *param;

        buf = bt_hci_cmd_create(VENDOR_OP, sizeof(*param));
        ...
        param->foo = 0x2c;
        param->bar = 0x01;
        param->baz = 0x01;
        ...
```

Even better if you also create defines for the magic values `0x2c`, `0x01` and `0x01`

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -670,6 +674,30 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
 }
 
+#if defined(CONFIG_BT_BLUENRG_ACI)
+static int hci_send_aci_config_data_controller_mode(void)
```
Ideally this kind of stuff should be completely on the driver side. Does HCI_Reset undo the effects of the command you're sending? If not, you could do the command sending in the drivers's open() callback.

cpriouzeau replied:
The command MUST be sent just after HCI_reset to have an effect on bluetooth component internal stack and it's not possible change the role of the internal stack before or after (much later on time) HCI_reset

jhedberg replied:
Is HCI_Reset; Vendor_command; HCI_Reset; also something that will not work? Will the second HCI_Reset throw the controller's internal host stack back on? If that's also not an option, we could add a new optional callback to HCI drivers (called e.g. "setup") that's called after the host has sent the first HCI_Reset. That way the hardware-specific init would still stay inside the HCI driver.

cpriouzeau replied:
no, it's only HCI_Reset; Vendor_command.

jhedberg replied:
Ok, in that case I think the most reasonable thing to do is an optional driver->setup() callback after hci_core.c has sent HCI_Reset. Btw, where can we find reference documentation to this vendor command, its parameters, and how the command relates to HCI_Reset? Is that information available somewhere publicly?

We're planning to add a bt_disable() to accompany bt_enable() at some point in the future, and that will most likely involve calling HCI_Reset again (when bt_enable() is called a second time), so we need to make sure that no behavior is broken.

jhedberg replied:
Ok, the way you've cleaned this up with appropriate defines and a parameter struct looks good. However, it'd still be good to move all of it over to `drivers/bluetooth/hci/spi.c` with the help of a new optional HCI driver callback. I've got a patch ready for this, but am a little hesitant since I'm not sure how easily this will be reusable for other vendor extensions (e.g. the exact order of HCI_Reset vs vendor commands), and how we should then deal with other drivers that require a different order.

Was it so that HCI_Reset essentially resets all vendor settings as well? Will it be possible to do HCI_Reset; vendor_write_config; HCI_Reset; vendor_write_config? This is important information for introducing a bt_disable() API (something that may happen already for Zephyr 1.13)

cpriouzeau replied:
@jhedberg I have made the test for HCI_Reset; vendor_write_config; HCI_Reset; vendor_write_config  and it doesn't work, my internal blueNRG stack are not on a stable state after this list of command. With my test with ipsp sample, only HCI_Reset; vendor_write_config (make one time) work correctly to perform the ipsp use cases.

jhedberg reviewed (changes requested):
Sorry for not looking at this earlier - I thought the PR only involved touching hardware (or HCI driver) specific files. I now see that hci_core.c is touched as well, and there are a few things that would be good to get cleaned up there.

SebastianBoe commented on arch/CMakeLists.txt:
```diff
@@ -1,4 +1,2 @@
-add_definitions(-D__ZEPHYR_SUPERVISOR__)
-
```
Is this intentional?

cpriouzeau replied:
Sorry, I have pushed with your revert of #7567, it's an error of my part, I will push again without this patch

cpriouzeau commented:
@SebastianBoe Sorry, I have pushed my new version of commit with the content of #7567, it's an error of my part, I have pushed again without this changes.

cpriouzeau commented:
@jhedberg Any feedback about the last proposal ?

jhedberg commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -460,6 +501,9 @@ static int bt_spi_open(void)
 	.name		= "BT SPI",
 	.bus		= BT_HCI_DRIVER_BUS_SPI,
+#if defined(CONFIG_BT_BLUENRG_ACI)
+	.quirks		= BT_QUIRK_NO_RESET,
```
When this quirk is set, the assumption is that the HCI driver takes care of HCI_Reset before open() returns. I don't see that in your patch anywhere? Also, you could add a bit more descriptive commit message (explaining why this all is needed).

cpriouzeau replied:
I update to commit message to explain it. At least with my test I have seen that when BlueNRG are initialised, it not necessary to make a 'HCI_reset', it's why I just force the hardware on controller mode after it's initialised.

jhedberg replied:
Yep, looks good now

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -85,12 +85,20 @@ config BT_SPI_DEV_NAME
+	  Enable support for devices compatible with the BlueNRG Bluetooth
+	  Stack. Current driver supports: ST BLUENRG-MS.
+
+
```
Redundant empty line here

jhedberg replied:
Looks like there's one in the commit message as well (which github doesn't let me do inline commenting on). If these minor issues are fixed I'd be happy to approve this PR.

cpriouzeau replied:
@jhedberg Modification done, Thanks for the review

cpriouzeau committed:
drivers/clock_control: stm32: fix HCLK calculation when using MSI

MSI clock signal can be selected in several ranges.
These ranges should be taken into account for calculating its
frequency and hence global system frequency.
Provide a Kconfig option to set MSI range.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>

cpriouzeau committed:
bt: hci driver over spi: Configure BlueNRG-MS in controller mode

Configure BlueNRG in controller w/o host.
By disabling 'HCI_reset' via the quirk, we can request to
BluesNRG hardware to be on controller mode just after it is ready.
By the way it's no necessary to manage a 'HCI_reset' when BluesNRG are
initialised (and on controler mode).

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

carlescufi commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -194,12 +217,65 @@ static void bt_spi_release_cs(void)
 	gpio_pin_write(cs_dev, GPIO_CS_PIN, 1);
 }
+
+static bool bt_spi_irq_pin_high(void)
```
You don't need prefixes (`bt_spi_` for static functions), can you remove that?

cpriouzeau replied:
@carlescufi are you sure because the other functions like bt_spi_configure_cs, bt_spi_kick_cs, bt_spi_release_cs have the bt_spi_ prefix ? and are not introduce by me. If I change the name of my new function, do I need to change this function name also ?

cpriouzeau replied:
@carlescufi I have pushed the new version of patch with this modification. Thanks

carlescufi commented:
@cpriouzeau just a nitpick, rest looks good

cpriouzeau committed:
bt: hci driver over spi: BlueNRG-MS read until IRQ pin goes low

According to BlueNRG-MS ref manual, SPI master shoul read  from
controller as long as IRQ pin is high.
Introduce a new loop within spi_rx_thread to do so. Since
it is now possible to enter read loop with nothing to read, set
a max attempts limit to be sure to exit the loop.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Christophe Priouzeau <christophe.priouzeau@st.com>

carlescufi reviewed (approved):
Thank you!

erwango commented:
recheck

cpriouzeau commented:
recheck

carlescufi commented:
recheck
